### DISCUSSION TITLE
Automatically deploy flask app didalam ec2.

### DISCUSSION THREAD

**Discussion by slawalata** [awsugid/forum#26] [General]
Salam,

- Untuk mendeploy _flask web app_ kedalam _ec2_ instance, maka saya membutuhkan bbrp hal untuk diinstall sebelumnya seperti:
1. Gunicorn
2. python dan pip
3. _Ngix_ dan setting untuk  _request routing_ dari _public ip _ kedalam _flask app_.

Dengan segala kerumitan diatas, jalan apa yang saya sebaiknya saya tempuh untuk mengotomatisasi proses diatas, apakah:
1. menggunakan boto3 untuk membuat ec2 instance (termasuk kedepan database) dan instance inisialisasi script untuk mempersiapkan environment (termasuk gunicorn).
2. atau ada cara lain untuk mempermudah mendeploy applikasi ini seperti aws pipeline builder. Mengingat kedepan mungkin architecturenya akan bertambah spt penambahan vpc dan database.

Urun rembuk, 

terima kasih atas kesediaanya untuk berdiskusi,

salam
slawalata

**Comment by rioastamal** [comment]
@slawalata, untuk mengotomasi pembuatan resource tidak hanya EC2 tapi juga database, VPC dan sebagainya bisa menggunakan Infrastructure as Code (IaC) tools. Beberapa opsi populer adalah Terraform, CloudFormation, Amazon CDK dan lainnya. IaC tools tersebut bisa dipadukan dengan CI/CI pipeline yang kamu sudah familiar misal GitHub Action, GitLab CI atau layanan dari AWS seperti CodeBuild dan CodePipeline.

Saran:
Daripada di-install langsung di EC2 lebih baik dipaket dulu menjadi container image sehingga aplikasi lebih portable dan bisa digunakan dibanyak layanan lainnya tanpa perlu modifikasi.

Ref:
- https://aws.amazon.com/blogs/opensource/automate-python-flask-deployment-to-the-aws-cloud/
- https://aws.amazon.com/tutorials/serve-a-flask-app/

**Reply by za** [reply]
> Saran:
Daripada di-install langsung di EC2 lebih baik dipaket dulu menjadi container image sehingga aplikasi lebih portable dan bisa digunakan dibanyak layanan lainnya tanpa perlu modifikasi.

Kak @rioastamal klo sudah dipaket jadi container, apakah nanti di-deploy di EC2 atau di ECS ya?

**Reply by rioastamal** [reply]
@za Ada banyak pilihan ketika aplikasi sudah menjadi container (diurutkan dari yang paling mudah, menurut saya):

- Lightsail container
- App Runner
- ECS
- Lambda
- EC2 (via Docker Engine dan sejenisnya)
- EKS (Kubernetes)
